crates/symbolicator-service/src/download/deny_list.rs

```diff
@@ -0,0 +1,224 @@
+use std::collections::VecDeque;
+use std::sync::{Arc, Mutex};
+use std::time::{Duration, SystemTime};
+
+use crate::caching::CacheError;
+use crate::config::Config;
+
+/// A record of a number of download failures in a given second.
+#[derive(Debug, Clone, Copy)]
+struct FailureCount {
+    /// The time at which the failures occurred, measured in milliseconds since the Unix Epoch.
+    timestamp: u64,
+    /// The number of failures.
+    failures: usize,
+}
+
+type CountedFailures = Arc<Mutex<VecDeque<FailureCount>>>;
+
+/// A structure that keeps track of download failures in a given time interval
+/// and puts hosts on a block list accordingly.
+///
+/// The logic works like this: if a host has at least `failure_threshold` download
+/// failures in a window of `time_window_millis` ms, it will be blocked for a duration of
+/// `block_time`.
+///
+/// Hosts included in `never_block` will never be blocked regardless of download_failures.
+#[derive(Clone, Debug)]
+pub(crate) struct HostDenyList {
+    time_window_millis: u64,
+    bucket_size_millis: u64,
+    failure_threshold: usize,
+    block_time: Duration,
+    never_block: Vec<String>,
```

Comment:
I know this was just copied from a different place in the code, but how long is this list? Any performance implication of this being a `Vec`?

Reply:
> Are you asking about the performance implications of the contains check?

Yes exactly, a set is what I would expect here.

Reply:
It's a good point, I made it a set now.